movement/io/validators.py

```diff
@@ -1,39 +1,348 @@
-            raise FileNotFoundError(error_msg)
-        return value
-
-    @field_validator("file_path")
-    def file_must_have_valid_suffix(cls, value):
-        if value.suffix not in (".h5", ".csv"):
-            error_msg = (
-                "Expected a file with pose estimation results from "
-                "DeepLabCut, in one of '.h5' or '.csv' formats. "
-                f"Received a file with suffix '{value.suffix}' instead."
-            )
-            logger.error(error_msg)
-            raise ValueError(error_msg)
-        return value
+    # Define class attributes
+    tracks_array: np.ndarray = field()
+    scores_array: Optional[np.ndarray] = field(default=None)
+    individual_names: Optional[List[str]] = field(
+        default=None,
+        converter=converters.optional(_list_of_str),
+    )
+    keypoint_names: Optional[List[str]] = field(
+        default=None,
+        converter=converters.optional(_list_of_str),
+    )
+    fps: Optional[float] = field(
+        default=None,
+        converter=converters.pipe(  # type: ignore
+            converters.optional(float), _set_fps_to_none_if_invalid
+        ),
+    )
+
+    # Add validators
+    @tracks_array.validator
+    def _validate_tracks_array(self, attribute, value):
+        _ensure_type_ndarray(value)
+        if value.ndim != 4:
+            log_and_raise_error(
+                ValueError,
+                f"Expected `{attribute}` to have 4 dimensions, "
+                f"but got {value.ndim}.",
+            )
+        if value.shape[-1] not in [2, 3]:
+            log_and_raise_error(
+                ValueError,
+                f"Expected `{attribute}` to have 2 or 3 spatial dimensions, "
+                f"but got {value.shape[-1]}.",
+            )
+
+    @scores_array.validator
+    def _validate_scores_array(self, attribute, value):
+        if value is not None:
+            _ensure_type_ndarray(value)
+            if value.shape != self.tracks_array.shape[:-1]:
+                log_and_raise_error(
+                    ValueError,
+                    f"Expected `{attribute}` to have shape "
+                    f"{self.tracks_array.shape[:-1]}, but got {value.shape}.",
+                )
+
+    @individual_names.validator
+    def _validate_individual_names(self, attribute, value):
+        if (value is not None) and (len(value) != self.tracks_array.shape[1]):
+            log_and_raise_error(
+                ValueError,
+                f"Expected {self.tracks_array.shape[1]} `{attribute}`, "
+                f"but got {len(value)}.",
+            )
+
+    @keypoint_names.validator
+    def _validate_keypoint_names(self, attribute, value):
+        if (value is not None) and (len(value) != self.tracks_array.shape[2]):
+            log_and_raise_error(
+                ValueError,
+                f"Expected {self.tracks_array.shape[2]} `{attribute}`, "
+                f"but got {len(value)}.",
+            )
+
+    def __attrs_post_init__(self):
```

Comment:
I thought this was the easiest way to do it, since the default values depend on the shape of tracks_array, which therefore needs to be accessed via "self.tracks_array". The "self" is only populated after initialisation.

I has also considered doing this during initialisation, usings the `attrs` [default decorator](https://www.attrs.org/en/23.1.0/init.html#defaults). However, there is a catch:

> __Watning__
> Please note that the decorator based defaults have one gotcha: they are executed when the attribute is set, that means depending on the order of attributes, the self object may not be fully initialized when they're called.

> Therefore you should use self as little as possible.

> Even the smartest of us can get confused by what happens if you pass partially initialized objects around.

In theory, since tracks_array is initialised first, it should be available to be used by the defaults of other parameters. But I decided I didn't want to play with partially initialised objects.

Reply:
Alles klar. Thanks for the clarification!